### PR TITLE
[FIX] Rectify: Evidence-Bound Injected Instruction Policy

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

**This PR implements PART A ONLY.** Part B was split into issue #4375, which must be sequenced after #4369.

`CODEX_INTAKE_DISCIPLINE_DIGEST` is a 790-byte, six-rule free-prose string at
`src/autoskillit/core/types/_type_constants.py:170-197`, injected into every Codex session through
five delivery surfaces and into no Claude session. Two of its six rules have a cited basis; four do
not. Rule 2 — *"Never read a file end-to-end"* — contradicts both OpenAI Codex's shipped
`base_instructions` (verified verbatim from a first-party rollout) and the sibling
`OUTPUT_DISCIPLINE_DIGEST` concatenated into the *same* injected string two bytes away.

Rewording rule 2 is a one-line diff, and the report's own evidence says it will not hold: between
one in four and one in two closed issues here is later contradicted by a recurrence, and the
dominant cause is aiming at the wrong layer. The wrong layer here is *prose*. The right layer is
**the absence of any structure that could represent, and therefore check, a relationship between
two instruction texts**.

Part A replaces the free-prose constant with an **evidence-bound rule registry** — each rule
carries a machine-checkable `basis`, `evidence`, `evidence_anchor`, `subject`, `exception` and
`path_classes` — and renders the wire text from it. Six guards then make the bug class fail loudly:

| Guard | Bug class it makes impossible |
|---|---|
| G1 — no absolute imperative without a declared exception | *"An unconditional imperative correct in its original context and wrong in a context nobody re-scoped it for"* — the report's central finding, and the generative cause of #4351, #4265 and #4373 |
| G2 — every rule's `basis` resolves to a live capability value, an existing ADR, or a cited issue/doc containing the rule's own anchor | Unmeasured prose (R1's `2`, R4's `2`) frozen into a contract by a substring test |
| G3 — no rule may name a path class the `recipe_read_guard` denies | An unexecutable carve-out — the exact trap the investigation flagged for recipe files |
| G4 — the *composed* suffix must round-trip through `tomllib` | The real gap at `codex.py:847-849`, which inspects only the agent body and never the appended suffix |
| G5 — declared byte budget on the digest and on the composed suffix | Unbudgeted growth in an always-on injection replicated 12× per session |
| G6 — co-injected policy texts must claim disjoint subjects | The `OUTPUT_DISCIPLINE_DIGEST` ↔ intake-digest contradiction, with no reference to upstream needed |

Behavioural verification needs **zero new plumbing**: the digest header carries its own version at
the wire (`Context Intake Discipline v2:`), and `sessions.jsonl` already stores `codex_log`, a
direct path to each Codex session's rollout. Bumping the version to 2 *is* the cohort key. Part A
ships a reproducible measurement script so re-measuring the 54.6% repeat-read rate is a command,
not archaeology.

## Scope Note

This PR is **Part A only** of a two-part rectification. Part B — the remainder of the
rectification — was split out into issue #4375, which must be sequenced **after** #4369.

Closes #4351

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/fix-4351-intake-digest-20260726-144540-308354/.autoskillit/temp/rectify/rectify_codex_intake_discipline_digest_2026-07-26_154500_part_a.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

### CODE REVIEW

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: changes_requested

File: scripts/measure_codex_read_repetition.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    worst_overall.sort(key=lambda row: -row[2])
+    return {
+        "cohorts": cohorts,
+        "worst_offenders": worst_overall[:20],
+        "unclassified_record_count": total_unclassified,
+    }
+
+
+def _find_rollouts(log_root: Path, since: str | None, until: str | None) -> list[Path]:
+    paths = sorted(log_root.glob("*/*/*/rollout-*.jsonl"))
```
Comment (Trecek) [inline scripts/measure_codex_read_repetition.py]: [critical] bugs: _find_rollouts globs log_root.glob("*/*/*/rollout-*.jsonl") (three wildcard directory levels), but the real on-disk Codex rollout layout is only two levels deep: codex-sessions/YYYY/MM/rollout-*.jsonl (confirmed by tests/execution/backends/test_codex_session_storage.py, e.g. Path("2026/07/rollout-resume.jsonl"), and by _codex_session_storage.py's promotion logic). Against a real log directory this glob matches zero files, so the tool silently reports rollouts_scanned=0 and measures nothing, with no error surfaced.

File: scripts/measure_codex_read_repetition.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+# The rarer custom_tool_call/exec shape embeds the command in a JS-template
+# string: tools.exec_command({cmd:"..."}).
+_CUSTOM_TOOL_CALL_CMD_PAT = re.compile(
+    r"tools\.exec_command\(\{\s*cmd\s*:\s*[\"']((?:[^\"'\\]|\\.)*)[\"']"
+)
+
+_VERSION_V2_MARKER = "Context Intake Discipline v2:"
+_VERSION_V1_MARKER = "Context Intake Discipline v1:"
+
+_ROLLOUT_DATE_PAT = re.compile(r"(\d{4})/(\d{2})/(\d{2})/rollout-")
```
Comment (Trecek) [inline scripts/measure_codex_read_repetition.py]: [critical] bugs: _ROLLOUT_DATE_PAT assumes a YYYY/MM/DD/rollout- three-component date path, but real rollout paths are YYYY/MM/rollout-*.jsonl (two components, no day). Even after fixing the glob depth at L222, this pattern will never match, so --since/--until date filtering silently falls through to "include everything" instead of actually filtering.

File: scripts/measure_codex_read_repetition.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    """Return True when cmd's *leading* command (before any pipe) is a bounded file read."""
+    return any(pattern.search(cmd) for pattern in _BOUNDED_PATTERNS)
+
+
+def extract_target_path(cmd: str) -> str | None:
+    """Best-effort extraction of the file path a bounded-read command targets."""
+    m = re.search(r"sed\s+-n\s+'[^']*'\s+([^\s|;]+)", cmd)
+    if m:
+        return m.group(1)
+    m = re.search(r"\brg\s+[^|;]*?(-n\b|--line-number\b)[^|;]*", cmd)
```
Comment (Trecek) [inline scripts/measure_codex_read_repetition.py]: [critical] bugs: The rg path-extraction regex r"\brg\s+[^|;]*?(-n\b|--line-number\b)[^|;]*" stops at the first literal | character, including one inside a quoted ripgrep pattern. This repo's own AGENTS.md documents `rg -n 'foo|bar'` (POSIX ERE alternation) as the idiomatic search invocation; for that command shape, extract_target_path returns a bogus token extracted from inside the pattern (e.g. "foo") instead of the actual file path, silently corrupting the repeat-read grouping key.

File: scripts/measure_codex_read_repetition.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    """Return (exec_commands, unclassified_count) for one rollout JSONL file.
+
+    Malformed JSONL lines are skipped, never fatal. A record that is
+    exec-shaped (function_call/exec_command or custom_tool_call/exec) but
+    whose command could not be extracted increments the unclassified count
+    instead of silently vanishing.
+    """
+    commands: list[tuple[int, str]] = []
+    unclassified = 0
+    with rollout_path.open("r", encoding="utf-8") as f:
```
Comment (Trecek) [inline scripts/measure_codex_read_repetition.py:120]: [warning] defense: In classify_rollout_records, a json.JSONDecodeError on a raw JSONL line is swallowed with a bare `continue` and no counter increment, unlike the exec-shaped-but-unextractable branch below it (which increments `unclassified`). A corrupted line that would have been an exec_command record vanishes silently instead of being surfaced via the unclassified count.

File: scripts/measure_codex_read_repetition.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    try:
+        parsed = json.loads(raw_args)
+    except (json.JSONDecodeError, TypeError):
+        return None
+    return parsed.get("cmd") if isinstance(parsed, dict) else None
+
+
+def classify_policy_cohort(rollout_path: Path) -> str:
+    """Return 'v2', 'v1', or 'none' from the intake-discipline header seen on the wire."""
+    text = rollout_path.read_text(encoding="utf-8", errors="ignore")
```
Comment (Trecek) [inline scripts/measure_codex_read_repetition.py]: [warning] defense: classify_policy_cohort() calls rollout_path.read_text(...) with no exception handling, invoked once per rollout file in a loop over potentially thousands of files. A single unreadable file raises an uncaught OSError/UnicodeError that aborts the entire batch measurement with no indication of which file failed, unlike the per-line JSONDecodeError handling just above it.

File: scripts/measure_codex_read_repetition.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+def _find_rollouts(log_root: Path, since: str | None, until: str | None) -> list[Path]:
+    paths = sorted(log_root.glob("*/*/rollout-*.jsonl"))
+    if since is None and until is None:
+        return paths
+    filtered = []
+    for p in paths:
+        m = _ROLLOUT_DATE_PAT.search(p.as_posix())
+        date = "-".join(m.groups()) if m else None
+        if date is not None:
+            if since is not None and date < since:
```
Comment (Trecek) [inline scripts/measure_codex_read_repetition.py]: [critical] bugs: `_find_rollouts` compares the month-precision date string extracted by `_ROLLOUT_DATE_PAT` (`"YYYY-MM"`, since the real on-disk layout has no day component) against `--since`/`--until` via plain Python string comparison — but the argparse help text for both flags documents day-precision examples (`"2026-07-18"`). For any day-precision `--since` value in the target month, `"YYYY-MM" < "YYYY-MM-DD"` is always `True` (a string always sorts less than a longer string it prefixes), so `--since 2026-07-18` silently excludes the *entire* month of July, including rollouts from July 19-31 that should pass the filter. Symmetrically, `"YYYY-MM" > "YYYY-MM-DD"` is always `False`, so `--until 2026-07-18` silently *includes* the full month rather than bounding it at the 18th. This is a real, demonstrable regression introduced by this exact fix commit (verified directly: `"2026-07" < "2026-07-18"` → `True` in Python) — before this diff, the date pattern never matched real 2-level paths at all, so filtering was a no-op; now it matches but at the wrong precision for the documented CLI usage. `test_find_rollouts_date_filter_actually_filters` only exercises month-precision since/until values (`"2026-07"`), which masks this exact defect. Suggested fix: normalize `since`/`until` to `YYYY-MM` precision before comparing (e.g. `since[:7]`), or validate/reject day-precision input and correct the help text to require month-only input.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review (round 2) — Verdict: changes_requested

**Review by Trecek — COMMENTED** [review]
AutoSkillit review (round 3) — Verdict: approved (posted as COMMENT, not APPROVE — GitHub disallows self-approval on this self-authored PR). Verified the round-2 day-precision date-filter fix (commit 7d59e278e) is correct and complete: manually traced `_date_within_bound` against all boundary cases and bisected the rewritten test against pre-fix logic to confirm it genuinely pins the regression, corroborated by two independent audit subagents. Round-1 fixes (97d8e8b84) remain intact; the swallowed-JSONDecodeError item remains the deliberate DISCUSS deferral, correctly left unaddressed. No critical/warning findings — only two info-level test-coverage suggestions (non-blocking). task test-check: 30702 passed, 0 failures (an initial run hit an unrelated /dev/shm xdist-contention flake, traced and confirmed not caused by this PR; clean re-run passed). pre-commit run --all-files: clean. This PR implements Part A of #4351 only; Part B is tracked separately as #4375.

File: scripts/measure_codex_read_repetition.py (hunk 1 above)
Comment (Trecek) [reply scripts/measure_codex_read_repetition.py]: Fixed in `a44ae9173` (branch commit `97d8e8b84`, squashed on merge). The glob is now `log_root.glob("*/*/rollout-*.jsonl")` at `scripts/measure_codex_read_repetition.py:252` — two wildcard directory levels (`YYYY/MM/`), matching the real on-disk layout rather than three.

File: scripts/measure_codex_read_repetition.py (hunk 2 above)
Comment (Trecek) [reply scripts/measure_codex_read_repetition.py]: Fixed in `a44ae9173` (branch commit `7d59e278e`, squashed on merge). `_ROLLOUT_DATE_PAT` no longer parses the directory path. It is now `re.compile(r"rollout-(\d{4}-\d{2}-\d{2})T")` at line 75 and is matched against `p.name`, so the date is derived at day precision from the filename and does not depend on directory depth at all.

File: scripts/measure_codex_read_repetition.py (hunk 3 above)
Comment (Trecek) [reply scripts/measure_codex_read_repetition.py]: Fixed in `a44ae9173` (branch commit `97d8e8b84`, squashed on merge). The extraction regex is now built from `_QUOTED_OR_UNPIPED` (line 89), which is alternation-safe and no longer truncates at the first literal `|`. A leading-command guard `_LEADING_RG_N_PAT` (line 45) was added alongside it.

File: scripts/measure_codex_read_repetition.py (hunk 4 above)
Comment (Trecek) [reply scripts/measure_codex_read_repetition.py:120]: Adjudicated **DISCUSS — deliberately not changed**, and recorded as such in the programme handoff rather than silently dropped.

Counting the swallowed `json.JSONDecodeError` into `unclassified` would flip `test_unparseable_records_are_skipped_not_fatal`, which asserts `unclassified == 0` on purpose. It would also conflate two different things: a *record I could not parse* and a *read I could not classify*. `unclassified` means the latter.

If this visibility is wanted later, the correct shape is a distinct `skipped_unparseable` counter reported alongside `unclassified`, not a merge of the two. Judged not worth its own ticket at the time; the reasoning is preserved in the handoff so it is not re-litigated from scratch.

File: scripts/measure_codex_read_repetition.py (hunk 5 above)
Comment (Trecek) [reply scripts/measure_codex_read_repetition.py]: Fixed in `a44ae9173`. `classify_policy_cohort()` now wraps the read:

```python
try:
    text = rollout_path.read_text(encoding="utf-8", errors="ignore")
except OSError:
    return "none"
```

with a docstring stating the intent explicitly — one unreadable file in a batch of thousands must not abort the whole measurement run.

File: scripts/measure_codex_read_repetition.py (hunk 6 above)
Comment (Trecek) [reply scripts/measure_codex_read_repetition.py]: Fixed in `a44ae9173` (branch commit `7d59e278e`, squashed on merge) by the same change that resolved the `_ROLLOUT_DATE_PAT` thread. The comparison is no longer month-precision: the date is extracted at day precision from the filename, so `_date_within_bound` compares like with like.
